GDJS/Runtime/howler-sound-manager/howler-sound-manager.js

```diff
@@ -14,87 +15,274 @@
  * @memberof gdjs
  * @class HowlerSound
  */
-gdjs.HowlerSound = function (o) {
-  Howl.call(this, o);
-  this._paused = false;
-  this._stopped = true;
-  this._canBeDestroyed = false;
-  this._rate = o.rate || 1;
-
-  //Add custom events listener to keep
-  //track of the sound status.
-  var that = this;
-  this.on('end', function () {
-    if (!that.loop()) {
-      that._canBeDestroyed = true;
-      that._paused = false;
-      that._stopped = true;
+gdjs.HowlerSound = class HowlerSound {
+  /**
+   * The ID of the played sound.
+   * @type {?number}
+   * @private
+   */
+  _id = null;
+
+  /**
+   * The Howl passed to the constructor.
+   * It defines the sound file that is being played.
+   */
+  _howl = null;
+
+  /**
+   * Internal variable to check if the sound is still loading.
+   * A loading sound will count as played.
+   * @type {boolean}
+   * @private
+   */
+  _loading = false;
+
+  /**
+   * An internal variable for tracking if the sound can be destroyed.
+   * This is used to remove a free sound when it finishes.
+   * @type {boolean}
+   * @private
+   */
+  _canBeDestroyed = false;
+
+  constructor(howl) {
+    this._howl = howl;
+  }
+
+  /**
+   * Begins playback of the sound.
+   * @returns {HowlerSound} The current instance for chaining.
+   */
+  play() {
+    if (this._howl._state === 'loaded') {
+      this._id = this._howl.play(this._id === null ? undefined : this._id);
+    } else {
+      this._loading = true;
+      // In case the howl didn't finsih loading, start the sound once it finishes loading.
+      this._howl.once('load', () => {
+        this._loading = false;
+        this.play();
+      });
+      // Make sure it is loading
+      this._howl.load();
     }
-  });
-  this.on('playerror', function (id, error) {
-    console.error(
-      "Can't play a sound, considering it as stopped. Error is:",
-      error
+
+    return this;
+  }
+
+  /**
+   * Sets the current sound as a free sound.
+   * This will set up a listener to mark the sound as destroyable
+   * after it ended.
+   * This has to be done separately as it could lead to a memory leak otherwise.
```

Comment:
Can you explain what do you mean by "this has to be done separately"?

Reply:
If we set up the end event on a non free sound, it may never be called (for example of it gets paused and never resumed). This would cause a leak, because the listener would never be removed, etc. On free sounds, we know that no interaction from outside can do that as they are not accessible, so we know it will reach its end at some point.